### PR TITLE
Fix #4630: Have order on msgids in sphinx.pot deterministic

### LINKED ISSUE #4630: Order on msgids included in sphinx.pot is not deterministic
Subject: Order on msgids in sphinx.pot has difference even if its content has no essential changes

### Problem
- When .pot files are managed with VCS (such as Git), difference without essential changes bother project manager.

#### Procedure to reproduce the problem
```
git clone https://github.com/python/cpython
cd cpython/Doc
make build ALLSPHINXOPTS="-E -b gettext -D gettext_compact=0 -d build/.doctrees . locales/pot"
```

#### Error logs / results
- https://github.com/python/python-docs-ja/commit/e23b4d3b310353073f1ba60b08160d8da09425db#diff-45573348f1c20fa58a6bb5ae4fbeb851

#### Expected results
Treated as "No difference".
More concretely, msgids in sphinx.pot have deterministic order.

### Reproducible project / your project
- https://github.com/python/cpython (.pot files are genereted from this project)

### Environment info
- OS: Linux/Travis CI
- Python version: 3.6.3
- Sphinx version: 1.7.0

### PR DESCRIPTION
**Pull request by cocoatomo**
ref. #4630 

### Feature or Bugfix
- Bugfix

### Purpose
- Have order on msgids in sphinx.pot deterministic

### Detail
- Have order on msgids in sphinx.pot deterministic

### CODE REVIEW

File: sphinx/builders/gettext.py [hunk 1]
```diff
@@ -237,7 +237,8 @@ def _collect_templates(self):
 
     def _extract_from_template(self):
         # type: () -> None
-        files = self._collect_templates()
+        files = list(self._collect_templates())
+        files.sort()
```
Comment (tk0miya) [inline sphinx/builders/gettext.py:240]: Note: I think this is responsibility of `_collect_templates()`. On the other hand, a compatibility is another topic of maintenance. So I'd like to go with as is.

**Review by tk0miya — CHANGES_REQUESTED** [review]
Could you rebase to 1.7 please?
And please add a testcase if possible.

File: sphinx/builders/gettext.py [hunk 2]
```diff
@@ -68,8 +68,8 @@ def __init__(self):
         # type: () -> None
         self.messages = []  # type: List[unicode]
                             # retain insertion order, a la OrderedDict
-        self.metadata = {}  # type: Dict[unicode, List[Tuple[unicode, int, unicode]]]
-                            # msgid -> file, line, uid
+        self.metadata = OrderedDict()  # type: Dict[unicode, List[Tuple[unicode, int, unicode]]]
```
Comment (cocoatomo) [inline sphinx/builders/gettext.py]: @tk0miya flake8 reports E501 line too long. Do you have any workaround for this error?
ref. https://travis-ci.org/sphinx-doc/sphinx/jobs/342938730#L495

Comment (cocoatomo) [reply sphinx/builders/gettext.py]: Could I define a type alias for the type of metadata in the body of `if TYPE_CHECKING:` statement,
and use that type alias here?

Comment (tk0miya) [reply sphinx/builders/gettext.py]: Please mark as `# NOQA` at tail.

File: tests/test_build_gettext.py [hunk 3]
```diff
@@ -165,3 +165,18 @@ def test_gettext_template(app):
     result = (app.outdir / 'sphinx.pot').text(encoding='utf-8')
     assert "Welcome" in result
     assert "Sphinx %(version)s" in result
+
+
+@pytest.mark.sphinx('gettext', testroot='intl-template')
```
Comment (tk0miya) [inline tests/test_build_gettext.py]: Could you rename this to `test-gettext-template`?
This is a bug of gettext builder (extracting part), not internationalization (applying part).

**Review by tk0miya — CHANGES_REQUESTED** [review]
LGTM with nits

**Comment by cocoatomo** [comment]
Just finished fixing pointed issues.

**Review by tk0miya — APPROVED** [review]

**Comment by tk0miya** [comment]
Merged!
Thank you for your contribution!

**Pull request merged**